lib/rust/frp/src/io/timer/timeout.rs

```diff
@@ -0,0 +1,102 @@
+/// started, an `on_expired` event will be emitted at most once after the specified time provided
+/// during start. If the timer is cancelled before it expires, no event will be emitted during that
+/// run. Starting the timer again while it is already running is treated as a restart - previous run
+/// is canceled.
+///
+/// in `start`:       5ms--------20ms--3ms-------3ms-------
+/// in `cancel`:      ----------------------------x----x---
+/// out `on_expired`: -----x--------------x----------------
+///
+/// The timer is based on `setTimeout` browser API. That means there is no guarantee about the
+/// exact time the event will be emitted. It might be delayed if the browser event loop is busy. If
+/// you need to fire events periodically, prefer [`Interval`] timer in order to avoid unpredictable
+/// event trigger rate.
+#[derive(Clone, CloneRef, Debug)]
+pub struct Timeout {
+    /// Starts the timer immediately with provided timeout value, specified in integer
+    /// milliseconds. If the timer was already started, it is cancelled and restarted.
+    pub start:      frp::Any<i32>,
+    /// Stops the timer if it is running. No `on_expired` events will be emitted until it is
+    /// started again.
+    pub cancel:     frp::Any,
+    /// Emitted when timer expires. At most one event is emitted for each
+    /// start of the timer.
+    pub on_expired: frp::Stream<()>,
+    raw_timeout:    Rc<RawTimeout>,
+}
+
+impl Timeout {
+    /// Constructor. Timer is initially not started.
+    pub fn new(network: &frp::Network) -> Self {
+        frp::extend! { network
+            on_expired <- any_mut();
+        }
+
+        let closure: TimerClosure = Closure::new(f!(() on_expired.emit(())));
+        let raw_timeout = Rc::new(RawTimeout::new(closure));
+
+        frp::extend! { network
+            start  <- any_mut::<i32>();
+            cancel <- any_mut::<()>();
+            eval   start((timeout) raw_timeout.start(*timeout));
+            eval_  cancel(raw_timeout.cancel());
+        }
+        let on_expired = on_expired.into();
+        Self { on_expired, start, cancel, raw_timeout }
+    }
+}
+
+#[derive(Debug)]
+struct RawTimeout {
+    closure:      TimerClosure,
+    timer_handle: RefCell<Option<i32>>,
+}
+
+impl RawTimeout {
+    fn new(closure: TimerClosure) -> Self {
+        Self { closure, timer_handle: default() }
+    }
+
+    fn start(&self, time: i32) {
+        let js_func = self.closure.as_js_function();
+        let result = window.set_timeout_with_callback_and_timeout_and_arguments_0(js_func, time);
+        let handle = result.expect("setTimeout should not fail when provided with valid callback");
+
+        if let Some(old_handle) = self.timer_handle.borrow_mut().replace(handle) {
+            window.clear_timeout_with_handle(old_handle);
+        }
+    }
+
+    fn cancel(&self) {
+        if let Some(handle) = self.timer_handle.borrow_mut().take() {
+            window.clear_timeout_with_handle(handle);
+        }
```

Comment:
code duplication

Reply:
Same argument as before - i think that small visual duplication with the difference in `replace` vs `take` is acceptable and definitely less surprising than alternatives.